### PR TITLE
React docs part 5

### PR DESCRIPTION
**Pull request by BenAtEventbrite**
Contains docs on:

- Use of `render()`
- Use of `state`

### CODE REVIEW

File: react/README.md [hunk 1]
```diff
@@ -1259,34 +1267,338 @@ For more info on `dangerouslySetInnerHTML`, see: [Dangerously Set innerHTML](htt
...
+
+_Friendly reminder:_ If you want an **entire** component to be conditionally rendered, the component must return `null`. Returning `undefined` will be an error.
 
 **[⬆ back to top](#table-of-contents)**
 
 ## State
 
+There are two ways of maintaining data in a React component: props and state.
+
+The props are used by a component's parent to configure the component and are immutable within the component.
```
Comment (vincent-eb) [inline react/README.md]: nitpick: Might be clearer visually to separate the props and state descriptions with small headers or at least bolding. ie **Props**: "asdfasdf" **State**:"sfgdgfsdfg"

File: react/README.md [hunk 2]
```diff
@@ -1259,34 +1267,338 @@ For more info on `dangerouslySetInnerHTML`, see: [Dangerously Set innerHTML](htt
...
 **[⬆ back to top](#table-of-contents)**
 
 ### Defaulting from props
 
-Coming soon...
+In general, using props to generate state is an anti-pattern because it results in duplicate "sources of truth" (see [Props in getInitialState as an anti-pattern](https://facebook.github.io/react/tips/props-in-getInitialState-as-anti-pattern.html)). But if your props is properly named to indicate that it's only used as seed data for the component's internally-controlled state, it's no longer an anti-pattern.
+
+We tend to prefix these types of props with `default*` to match the `defaultValue` prop React uses for input elements. `initial*` is also a good prefix.
+
+When defaulting from props, using the declarative class field no longer works as the props are not available in static scope. The only option is to initialize within the `constructor`:
```
Comment (vincent-eb) [inline react/README.md]: I know that it is mentioned but maybe emphasize further that this style of initialization is only okay when utilizing default props. Previous example saying this is something that shouldn't be done may cause confusion.

File: react/README.md [hunk 3]
```diff
@@ -1259,34 +1267,338 @@ For more info on `dangerouslySetInnerHTML`, see: [Dangerously Set innerHTML](htt
...
+    constructor(props, context) {
+        super(prop, context);
+        this.state = {visible: props.visible};
+    }
+
+    // rest of the component
+}
+```
+
+In the "bad" example, both `props` and `state` have a property called `visible`, which is very confusing. Should you use `this.props.visible` or `this.state.visible`. The one in `props` cannot change, while the one in `state` can. Naming the prop `defaultVisible` (as shown in the "good" example) makes things clearer.
```
Comment (vincent-eb) [inline react/README.md]: I like this pattern! Haven't seen it before :D

File: react/README.md (hunk 1 above)
Comment (BenAtEventbrite) [inline react/README.md]: good idea!

File: react/README.md (hunk 2 above)
Comment (BenAtEventbrite) [inline react/README.md]: k, i'll add an additional statement at the end of this section

File: react/README.md (hunk 3 above)
Comment (BenAtEventbrite) [inline react/README.md]: wooo!

**Pull request merged**